SAM2_Tracking/utils.py

```diff
@@ -35,6 +36,211 @@ def read_config_yaml(config_path):
+            if len(value) > 1: # Lists with more than one value
+                trial_config[key] = value[i] # Assign the i-th value
+            else: # List with a single value
+                trial_config[key] = value[0] # Assign the first (only) value
+        else: # single non-list value
+            trial_config[key] = value
+    return trial_config
+
+def run_segmentation(config_file, device):
+    """
+    Runs SAM2 segmentation and mask propagation for one or more trial configurations.
+
+    This function reads a YAML configuration file and processes it to extract 
+    parameters for each trial. Each configuration parameter in the YAML file must 
+    either be a single value (applied to all trials) or a list of values (one per trial). 
+    For each trial, the function initializes a `SAM2FishSegmenter` object with the appropriate 
+    configuration and executes the segmentation and propagation process.
+
+    Parameters
+    ----------
+    config_file : str
+        Path to the YAML configuration file containing all segmentation parameters.
+        Each parameter should either be a scalar (applied to all trials) or a list of values
+        (with one entry per trial).
+    device : torch.device 
+            A `torch.device` class specifying the device to use for `build_sam2_video_predictor`
+
+    Returns
+    -------
+    None
+        The function does not return anything explicitly. However, it saves a pickled 
+        dictionary of masks (as specified by `masks_dict_file` in each trial config) 
+        for each trial after segmentation and propagation.
+
+    Notes
+    -----
+    - The number of trials is determined by the length of list-type parameters in the 
+      configuration file. All such parameters must have the same length.
+    - The function uses `read_config_yaml`, `extract_config_lens`, and `get_trial_config` 
+      as helpers in utils to parse and manage configurations.
+    
+    Warnings
+    --------
+    If the configuration specifies only a single `masks_dict_file` name while running multiple 
+    trials, the output masks will be written to the same file, and results will be 
+    overwritten. To prevent this, provide a list of unique `masks_dict_file` names — 
+    one for each trial.
+    
+    Examples
+    --------
+    >>> run_segmentation("template_configs.yaml", device=torch.device("cuda"))
+    Processing Trial 0: Frames from ./data/frames1, Annotations from ./data/annotations1.npy, Masks saving to ./generated_frame_masks1.pkl
+    Processing Trial 1: Frames from ./data/frames2, Annotations from ./data/annotations2.npy, Masks saving to ./generated_frame_masks2.pkl
+    """
+    # Load the YAML configuration file
+    configs = read_config_yaml(config_file)
+    
+    # Retrieve trial count from the length of values provided for each configuration key
+    trial_count = extract_config_lens(configs)
+    print(f"There are {trial_count} trials provided for processing")
```

Comment:
```suggestion
    print(f"Running segmentation for {trial_count} trial(s)")
```
You don't have to chose my suggestion, but I think this statement should be a little more specific and the grammar corrected.